### PR TITLE
Introduce parAll

### LINKED ISSUE #119: Introduce parAll
Into the companion object of `IO`, introduce the following `parAll` method:

```scala
final def parAll[E, A, M[X] <: TraversableOnce[X]](
    as: M[IO[E, A]]
  )(implicit cbf: CanBuildFrom[M[IO[E, A]], A, M[A]]): IO[E, M[A]]
```

This method should run all values in parallel. For now, it can be build by using the `par` method on the `IO` class.

Clues as to how to build this (given the complex type signature) can be found by examining the implementation of `forkAll`, which uses a similar type signature.

### PR DESCRIPTION
**Pull request by sideeffffect**
Fixes #119

### CODE REVIEW

**Comment by sideeffffect** [comment]
Now the question is, should we abstract away the common bits in `forkAll`

as.foldLeft(point\[E, Fiber\[E, mutable.Builder\[A, M\[A\]\]\]\](**Fiber.point(cbf(as))**)) { (asFiberIO, aIO) =>
        asFiberIO.par(aIO.**fork**).map {
          case (asFiber, aFiber) => asFiber.**zipWith**(aFiber)(\_ += \_)
        }
      }
      .map(**\_.map**(\_.result))

and `parTraverse`

as.foldLeft(point\[E, mutable.Builder\[B, M\[B\]\]\](cbf(as))) { (bsIO, a) =>
        bsIO.par(**fn**(a)).map {
          case (bs, b) => bs += b
        }
      }
      .map(\_.result)

?

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 1]
```diff
@@ -880,24 +881,16 @@ object IO {
...
-      if (!as.hasNext)
-        ioref
-      else {
-        val a = as.next
-        parTraverse_rec(as, ioref.par(fn(a)).flatMap { case (ref, b) => ref.modify(b :: _) *> point(ref) })
+    as: M[A]
+  )(fn: A => IO[E, B])(implicit cbf: CanBuildFrom[M[A], B, M[B]]): IO[E, M[B]] =
+    as.foldLeft(point[E, mutable.Builder[B, M[B]]](cbf(as))) { (bsIO, a) =>
+        bsIO.par(fn(a)).map {
+          case (bs, b) => bs += b
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: I think builders are not thread-safe 🤔

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 2]
```diff
@@ -824,10 +823,12 @@ object IO {
   final def forkAll[E, A, M[X] <: TraversableOnce[X]](
     as: M[IO[E, A]]
   )(implicit cbf: CanBuildFrom[M[IO[E, A]], A, M[A]]): IO[E, Fiber[E, M[A]]] =
-    as.foldRight(IO.point[E, Fiber[E, mutable.Builder[A, M[A]]]](Fiber.point(cbf(as)))) {
-        case (a, as) => as.par(a.fork).map { case (as, a) => as.zipWith(a)(_ += _) }
+    as.foldLeft(point[E, Fiber[E, mutable.Builder[A, M[A]]]](Fiber.point(cbf(as)))) { (asFiberIO, aIO) =>
```
[the remark this answers is not in the document]
Comment (sideeffffect) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: fair enough, fixed

**Comment by sideeffffect** [comment]
@jdegoes but we're using `IO.point` not for side-effecty things, but for normal, referentially transparent, `cbf(as)`
just as the previous code did
or am I missing something?

**Comment by sideeffffect** [comment]
ok, `parAll` it is

**Comment by jdegoes** [comment]
@sideeffffect `cbf(as)` has an effect (it allocates new mutable memory). It's possible the previous code was wrong and should have used `IO.sync`.

**Comment by sideeffffect** [comment]
@jdegoes ok, `IO.sync`
thanks for the explanation

**Comment by sideeffffect** [comment]
Although, I must admit that I don't *feel* that allocating a memory (mutable or not) is considered a side effect and that `IO.point` wouldn't break anything. The reason is that `cbf(as)` always returns "the same" object.
But I'm not 100 % sure, and having `IO.sync` can't do any harm.

**Comment by jdegoes** [comment]
@sideeffffect Yes, it's a side effect. You can verify it's not referentially transparent because if you feed the reference to 2 `IO` functions, it has a different effect than replicating the effect (`cbf(as)`) twice and feeding each one separately to each of the two `IO` functions.

**Comment by sideeffffect** [comment]
Btw I'm also introducing `sync` for `Fiber`, so that it's

IO.**sync**\[E, Fiber\[E, mutable.Builder\[A, M\[A\]\]\]\](Fiber.**sync**(cbf(as)))

**Comment by jdegoes** [comment]
`sync` for `Fiber` is not necessary because you are putting the effectfully-built builder inside `IO`.

Honestly, code like this which uses mutable builder is not great. It's not trivial to mix mutable builders with FP-style safely. But unless Scala has an immutable builder, there's not much you can do.

**Comment by jmcardon** [comment]
@jdegoes would it be that horrible to just... make a `parAll` for `List` and `Vector`? I mean realistically, how many other data structures will you be folding on list this?

**Comment by sideeffffect** [comment]
`Fiber.sync` removed

**Comment by sideeffffect** [comment]
tests added :)

**Comment by jdegoes** [comment]
@sideeffffect Looks good! Thank you for your contribution! ❤️

**Pull request merged**